orttraining/orttraining/python/training/orttrainer.py

```diff
@@ -1202,8 +1219,83 @@ def load_state_dict(self, state_dict, strict=True):
+                to the user when load_checkpoint is called.
+            include_optimizer_states: boolean flag indicating whether or not to persist the optimizer states.
+                on load_checkpoint, only model states will be loaded if include_optimizer_states==True
+        """
+
+        # extract state_dict to be saved in the checkpoint
+        state_dict = self.state_dict()
+
+        # if user_dict is provided, serialize to bytes and convert to hex string.
+        # this helps in loading the types as they are given by the user since hdf5
+        # converts to numpy types otherwise
+        if bool(user_dict):
+            state_dict[_utils.state_dict_user_dict_key()] = _checkpoint_storage.to_serialized_hex(user_dict)
+
+        # if include_optimizer_states is False, only save the model states in the checkpoint file
+        if not include_optimizer_states:
+            if _utils.state_dict_optimizer_key() in state_dict:
+                del state_dict[_utils.state_dict_optimizer_key()]
+
+        _checkpoint_storage.save(state_dict, path)
+
+    def _aggregation_required(self, loaded_trainer_options):
+        """Checks if aggregation is required for the loading the state_dict into the ORTTrainer"""
+
+        # To load states in the backend, aggregation is required for every ZeRO checkpoint
+        return loaded_trainer_options['zero_stage'] > 0
+
+    def load_checkpoint(self, *paths, strict=True):
+        """Loads the saved checkpoint state dictionary into the ORTTrainer
+
+        Reads the saved checkpoint files specified by paths from disk and loads the state dictionary
+        onto the ORTTrainer.
+        Aggregates the checkpoint files if aggregation is required.
+
+        Args:
+            paths: one or more files represented as strings where the checkpoint is saved
+            strict: boolean flag to strictly enforce that the saved checkpoint state_dict
+                keys match the keys from ORTTrainer.state_dict
+        Returns:
+            dictionary that the user had saved when calling save_checkpoint
+        """
+        state_dict = {}
+
+        # check if aggregation is required
+        loaded_trainer_options = _checkpoint_storage.load(paths[0], key=_utils.state_dict_trainer_options_key())
+        if self._aggregation_required(loaded_trainer_options):
+            # if aggregation is required, aggregation logic must be run on the saved checkpoints
+            state_dict = checkpoint.aggregate_checkpoints(paths, pytorch_format=False)
+        else:
+            # if aggregation is not required, reorder the checkpoints in order of ascending rank,
+            # and load the checkpoint for the current ORTTrainer rank.
+            ordered_paths = checkpoint._order_paths(paths)
+            state_dict = _checkpoint_storage.load(ordered_paths[self.options.distributed.world_rank])
+            assert state_dict['trainer_options']['world_rank'] == (self.options.distributed.world_rank), \
```

Comment:
in the case of single node or pure dataparallel run, there will be only a single checkpoint file saved since all workers have the same state. Trying to load a single checkpoint will error out here without reason since we will be loading an arbitrary rank's checkpoint file into all ranks.

Reply:
this condition and ordering should be checked only if multiple checkpoint paths have been provided

Reply:
you're right. Let me remove this check. It is not required since we are going to aggregate zero anyway.

Reply:
other than removing the assert, we also need to account for when a single checkpoint is loaded into all ranks. We do not need to order or index by rank in that case.